src/gevent/_hub_primitives.py

```diff
@@ -101,69 +101,73 @@ def __init__(self, objects, hub, timeout, count):
         self._hub = hub
         self._waiter = MultipleWaiter(hub) # pylint:disable=undefined-variable
         self._switch = self._waiter.switch
-        self._timeout = timeout
-        self._objects = objects
-
-        self._timer = None
+        self._timer = (
+            self._hub.loop.timer(timeout, priority=-1)
+            if timeout is not None else None
+        )
         self._begun = False
 
-
-        # Even if we're only going to return 1 object,
-        # we must still rawlink() *all* of them, so that no
-        # matter which one finishes first we find it.
-        self._count = len(objects) if count is None else min(count, len(objects))
-
-
-    def __iter__(self):
-        # When we begin iterating, we begin the timer.
         # XXX: If iteration doesn't actually happen, we
         # could leave these links around!
-        if not self._begun:
-            self._begun = True
+        self._objects = set(objects)
```

Comment:
I don't think we can count on objects being hashable. Overall the simpler PR in #1288 is probably better.